File: app/src/main/java/org/sheedon/demo/MainActivity.java
package org.sheedon.demo;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;

import org.json.JSONException;
import org.json.JSONObject;
import org.sheedon.demo.factory.MqttClient;
import org.sheedon.mqtt.Call;
import org.sheedon.mqtt.Callback;
import org.sheedon.mqtt.Observable;
import org.sheedon.mqtt.OkMqttClient;
import org.sheedon.mqtt.Request;
import org.sheedon.mqtt.RequestBuilder;
import org.sheedon.mqtt.Response;
import org.sheedon.mqtt.ResponseBody;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        final OkMqttClient client = MqttClient.getInstance().getClient();

        Request request = new RequestBuilder()
                .backName("get_manager_list")
                .build();

        final Observable observable = client.newObservable(request);
        observable.subscribe(new Callback<Response>() {
            @Override
            public void onFailure(Throwable e) {
                System.out.println(e.getMessage());
            }

            @Override
            public void onResponse(Response response) {
                ResponseBody body = response.body();
                System.out.println(body == null?"":body.getBody());
                observable.cancel();
            }
        });

        findViewById(R.id.btn).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                JSONObject jsonObject = new JSONObject();
                try {
                    jsonObject.put("type", "get_manager_list");
                    jsonObject.put("upStartTime", "");
                } catch (JSONException e) {
                    e.printStackTrace();
                }

                Request request = new RequestBuilder()
                        .payload(jsonObject.toString())
                        .backName("get_manager_list")
                        .build();

                Call call = client.newCall(request);
                call.enqueue(new Callback<Response>() {
                    @Override
                    public void onFailure(Throwable e) {

                    }

                    @Override
                    public void onResponse(Response response) {

                    }
                });

//                MqttClient.getInstance().publish(jsonObject.toString(), "get_manager_list", new Callback() {
//                    @Override
//                    public void onFailure(Throwable e) {
//                        System.out.println(e.getMessage());
//                    }
//
//                    @Override
//                    public void onResponse(Response response) {
//                        ResponseBody body = response.body();
//                        System.out.println(body == null?"":body.getBody());
//                    }
//                });
            }
        });
    }
}

File: app/src/main/java/org/sheedon/app/MainActivity.kt
package org.sheedon.app

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.View
import org.json.JSONException
import org.json.JSONObject
import org.sheedon.app.factory.MqttClient
import org.sheedon.mqtt.*

class MainActivity : AppCompatActivity() {

    val client: OkMqttClient = MqttClient.getInstance().getClient()!!

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val request: Request = Request.Builder()
            .backTopic("get_manager_list")
            .topic("yh_classify/device/recyclable/data/yhkhs20181029046")
            .delaySecond(10)
            .build()

        val observable = client.newObservable(request)
        observable.enqueue(object :Callback{
            override fun onResponse(call: Call, response: Response) {
                Log.v("TAG", "response:${response.body}")
            }

            override fun onFailure(e: Throwable?) {
                Log.v("TAG", "e:$e")
            }

        })

    }

    fun onTouchClick(view: View) {

        val jsonObject = JSONObject()
        try {
            jsonObject.put("type", "get_manager_list")
            jsonObject.put("upStartTime", "")
        } catch (e: JSONException) {
            e.printStackTrace()
        }

        val request = Request.Builder()
            .delaySecond(100)
            .topic("yh_classify/device/recyclable/data/yhkhs20181029046")
            .backTopic("yh_classify/clouds/recyclable/cmd/yhkhs20181029046")
//            .keyword("get_manager_list")
            .data(jsonObject.toString())
            .build()

        val call = client.newCall(request)
        call.enqueue(object :Callback{

            override fun onResponse(call: Call, response: Response) {
                Log.v("TAG", "response:${response.body}")
            }

            override fun onFailure(e: Throwable?) {
                Log.v("TAG", "e:$e")
            }
        })

//        MqttClient.getInstance().getClient()?.subscribe(Topics.build(
//            "yh_classify/clouds/recyclable/cmd/yhkhs20181029046",
//            1
//        ))

    }
}